type-inference-in-readr-and-arrow.rmd

```diff
@@ -0,0 +1,271 @@
+the two.
+
+<p align="center">
+![](images/readr_hex.png){width=200px}
+![](images/arrow-logo_hex_black-txt_white-bg.png){width=200px}
+</p>
+
+```{r, include=FALSE}
+library(magrittr)
+library(arrow)
+library(readr)
+```
+
+## How does type inference work in readr?
+
+In readr, the first step is reading in a subset of the data (by default,
+1000 rows, though this can be customised). The type inference is done by
+a C++ function in readr called `collectorGuess()` which guesses types in
+the following order:
+
+-   Does the column contain 0 rows? If yes, return "character"
+
+-   Are all values missing? If yes, return "logical"
+
+-   Tries to parse column to each of these formats and returns the first
+    one it successfully parses:
+
+    -   Logical
+
+    -   Integer (though the default is to not look for these)
+
+    -   Double
+
+    -   Number (a special type which can remove characters from strings
+        representing numbers and then convert them to doubles)
+
+    -   Time
+
+    -   Date
+
+    -   Datetime
+
+    -   Character
+
+The ordering above in the parsing bullet point goes from most to least
+strict in terms of the conditions which have to be met to successfully
+parse an input as that data type. For example, for a column to be of
+logical type, it can only contain `TRUE`, `FALSE`, or `NA` values, which
+is why this is the most strict type, but all of the other types could be
+read in as character data, which is the least strict and why this is
+last in the order.
+
+## How does type inference work in arrow?
+
+In arrow, `read_csv_arrow()` handles CSV reading, and much of its
+interface has been designed to closely follow the excellent APIs of
+`readr::read_csv()` and `vroom::vroom()`. The intention here is that
+users can use parameter names they're familiar with from the
+aforementioned readers when using arrow, and get the same results. The
+underlying code is pretty different though, and type inference is done
+on the whole column rather than an initial subset.
+
+In addition, Arrow has a different set of possible data types compared
+to R; see [the Arrow docs](https://arrow.apache.org/docs/r/articles/arrow.html) for more
+information about the mapping between R data types and Arrow types.
+
+In [the Arrow docs](https://arrow.apache.org/docs/cpp/csv.html#data-types), we can see
+that types are inferred in this order:
+
+-   Null
+
+-   Int64
+
+-   Boolean
+
+-   Date32
+
+-   Timestamp (with seconds unit)
+
+-   Float64
+
+-   Dictionary<String> (if ConvertOptions::auto_dict_encode is true)
+
+-   Dictionary<Binary> (if ConvertOptions::auto_dict_encode is true)
+
+-   String
+
+-   Binary
+
+Note that if you use `arrow::read_csv_arrow()` with parameter
+`as_data_frame = TRUE` (the default), the Arrow data types are then
+converted to R data types.
+
+```{r}
+simple_data <- data.frame(x = c(1, 2, 3), y = c("a", "b", "c"), z = c(1.1, 2.2, 3.3))
+
+write.csv(simple_data, "simple_data.csv", row.names = FALSE)
```

Comment:
Did you think about `write_csv` here instead? I was thinking it might be a touch less confusing to just stay in readr.

Reply:
Good point, will do!